### PR TITLE
[Improve](array-type) update stream_load for array nested array

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes
Now We can not load array nested array to doris by stream load, because we use from_string() which is so hardcode 
so I add a from_json for complex type such as array,map, struct , to support in from_string , later we can support complex type nested complex type with stream load 
Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: regression-test/suites/load_p0/stream_load/test_array_nested_array_load.groovy [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (amorynan) [reply regression-test/suites/load_p0/stream_load/test_array_nested_array_load.groovy:1]: OK I will make more wrong cases

File: be/src/vec/data_types/data_type_array.cpp [hunk 2]
```diff
@@ -267,56 +239,16 @@ Status DataTypeArray::from_string(ReadBuffer& rb, IColumn* column) const {
...
+    // json parser
+    std::unique_ptr<simdjson::ondemand::parser> _ondemand_json_parser =
+            std::make_unique<simdjson::ondemand::parser>();
+    size_t _padded_size = rb.count() + simdjson::SIMDJSON_PADDING;
+    std::string _simdjson_ondemand_padding_buffer;
+    _simdjson_ondemand_padding_buffer.resize(_padded_size);
+    memcpy(&_simdjson_ondemand_padding_buffer.front(), rb.position(), rb.count());
+    simdjson::ondemand::document array_doc = _ondemand_json_parser->iterate(
+            std::string_view(_simdjson_ondemand_padding_buffer.data(), rb.count()), _padded_size);
+    auto value = array_doc.get_value();
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_array.cpp:251]: should check if it is array

File: be/src/vec/data_types/data_type_array.cpp [hunk 3]
```diff
@@ -188,77 +188,49 @@ std::string DataTypeArray::to_string(const IColumn& column, size_t row_num) cons
...
+    for (auto it = outer_array.begin(); it != outer_array.end(); ++it) {
+        Status st;
+        try {
+            if (is_complex_type(remove_nullable(nested))) {
+                simdjson::ondemand::value val;
+                (*it).get(val);
+                st = nested->from_json(val, &nested_null_col);
+            } else {
+                std::string_view sv = (*it).raw_json_token().value();
+                ReadBuffer nested_rb(const_cast<char*>(sv.data()), sv.size());
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_array.cpp:210]: it seems little confusing for me that some use `from_json` while others use `from_string`.

Comment (amorynan) [reply be/src/vec/data_types/data_type_array.cpp:210]: when in complex type, we always in function_cast to call data-type from_string(), but this is not suitable for nested complex type to deserliaze, so when we call from_string from complex type, just call from_json thats ok. and we should decrease using from_string which will not exists in the future

File: be/src/vec/data_types/data_type_array.cpp (hunk 2 above)
Comment (amorynan) [reply be/src/vec/data_types/data_type_array.cpp:251]: I think it should in from_json begin avoid other complex type to call from_json

**Comment by amorynan** [comment]
run buildall

File: be/src/vec/data_types/data_type_nullable.cpp [hunk 4]
```diff
@@ -80,10 +81,27 @@ void DataTypeNullable::to_string(const IColumn& column, size_t row_num,
     }
 }
 
+Status DataTypeNullable::from_json(simdjson::ondemand::value& json_value, IColumn* column) const {
+    DCHECK(is_complex_type(nested_data_type));
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_nullable.cpp:85]: why dcheck complex in DataTypeNullable?

File: be/src/vec/data_types/data_type_array.cpp [hunk 5]
```diff
@@ -188,77 +189,64 @@ std::string DataTypeArray::to_string(const IColumn& column, size_t row_num) cons
...
-    while (!rb.eof() && (*rb.position() != ',') && (rb.count() != 1 || *rb.position() != ']')) {
-        // invalid elements such as ["123" 456,"789" 777]
-        // correct elements such as ["123"    ,"789"    ]
-        if (has_quota && !isspace(*rb.position())) {
-            return false;
+    simdjson::ondemand::array outer_array = json_value.get_array();
+    auto* array_column = assert_cast<ColumnArray*>(column);
+    auto& offsets = array_column->get_offsets();
+    IColumn& nested_column = array_column->get_data();
+    DCHECK(nested_column.is_nullable());
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_array.cpp:201]: what if nested is not nullable?

File: be/src/vec/data_types/data_type_nullable.cpp (hunk 4 above)
Comment (eldenmoon) [reply be/src/vec/data_types/data_type_nullable.cpp:85]: what if it's not complex type, i think we should not DCHECK complex type here

File: be/src/vec/data_types/data_type_array.cpp [hunk 6]
```diff
@@ -188,77 +189,64 @@ std::string DataTypeArray::to_string(const IColumn& column, size_t row_num) cons
...
+                }
+            } else {
+                std::string_view sv = simdjson::trim((*it).raw_json_token().value());
+                if (is_string_nested) {
+                    StringRef sr(sv.data(), sv.size());
+                    StringRef del("\"");
+                    sv = simd::VStringFunctions::trim(sr, del);
+                }
+                ReadBuffer nested_rb(const_cast<char*>(sv.data()), sv.size());
+                st = nested->from_string(nested_rb, &nested_column);
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_array.cpp:227]: i don't think it's quite reasonable to call from_string here, `from_json` semantic is different from `from_string`

Comment (eldenmoon) [reply be/src/vec/data_types/data_type_array.cpp:227]: mix them could be miss leading

File: be/src/vec/data_types/data_type_array.cpp [hunk 7]
```diff
@@ -188,77 +189,64 @@ std::string DataTypeArray::to_string(const IColumn& column, size_t row_num) cons
...
+    auto* array_column = assert_cast<ColumnArray*>(column);
+    auto& offsets = array_column->get_offsets();
+    IColumn& nested_column = array_column->get_data();
+    DCHECK(nested_column.is_nullable());
+    auto& nested_null_col = reinterpret_cast<ColumnNullable&>(nested_column);
+    bool is_string_nested = is_string(remove_nullable(nested));
+    size_t element_num = 0;
+    for (auto it = outer_array.begin(); it != outer_array.end(); ++it) {
+        Status st;
+        try {
```
Comment (eldenmoon) [inline be/src/vec/data_types/data_type_array.cpp:207]: why not `try catch`outof for loop? could iterate array throw simd execption?

File: be/src/vec/data_types/data_type_nullable.cpp (hunk 4 above)
Comment (eldenmoon) [reply be/src/vec/data_types/data_type_nullable.cpp:85]: from_nested_json

File: be/src/vec/data_types/data_type_array.cpp (hunk 7 above)
Comment (amorynan) [reply be/src/vec/data_types/data_type_array.cpp:207]: done!

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Review by eldenmoon — DISMISSED** [review]
LGTM

**Review by xy720 — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run p0

**Pull request closed without merging**